OpenRA.Mods.Common/Traits/Transforms.cs

```diff
@@ -127,11 +132,23 @@ public void DeployTransform(bool queued)
 				Faction = faction
 			};
 
-			var makeAnimation = self.TraitOrDefault<WithMakeAnimation>();
+			var checkAndTransform = new CallFunc(() =>
+			{
+				// If the deploy fails, try to salvage the situation by reversing the unmake animation.
+				if (CanDeploy(false))
+					self.QueueActivity(transform);
+				else
+				{
+					// Pass null for the second parameter so we don't issue pointless BuildingComplete notifications.
+					if (makeAnimation != null)
+						makeAnimation.Forward(self, null);
+				}
+			});
+
 			if (makeAnimation != null)
```

Comment:
As mentioned in the earlier discussion you also want to check this *before* running the make animation to filter out the cases that will always fail (e.g. trying to deploy next to invalid cells).